lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+                    self._push_task(dep_pkg, False, 0, 0, STATUS_ADDED)
+
+                # Clear any persistent failure markings _unless_ they are
+                # associated with another process in this parallel build
+                # of the spec.
+                spack.store.db.clear_failure(dep, force=False)
+
+            # Push any missing compilers (if requested) as part of the
+            # package dependencies.
+            if install_compilers:
+                self._add_bootstrap_compilers(self, self.pkg)
+
+        if install_self and self.pkg.unique_id not in self.build_tasks:
+            # Now add the package itself, if appropriate
+            self._push_task(self.pkg, False, 0, 0, STATUS_ADDED)
+
+    def _install_task(self, task, **kwargs):
+        """
+        Perform the installation of the requested spec and or dependency
+        represented by the build task.
+
+        Args:
+            task (BuildTask): the installation build task for a package"""
+
+        cache_only = kwargs.get('cache_only', False)
+        dirty = kwargs.get('dirty', False)
+        fake = kwargs.get('fake', False)
+        install_source = kwargs.get('install_source', False)
+        keep_stage = kwargs.get('keep_stage', False)
+        skip_patch = kwargs.get('skip_patch', False)
+        tests = kwargs.get('tests', False)
+        use_cache = kwargs.get('use_cache', True)
+        verbose = kwargs.get('verbose', False)
+
+        pkg = task.pkg
+        explicit = pkg.unique_id == self.pkg.unique_id
+
+        tty.msg(get_install_msg(pkg.unique_id))
+        task.start = task.start if task.start else time.time()
+        task.status = STATUS_INSTALLING
+
+        # Use the binary cache if requested
+        if use_cache:
+            if _installed_from_cache(pkg, cache_only, explicit):
+                self._update_installed(task)
+                return
+
+        pkg.run_tests = (tests is True or tests and pkg.name in tests)
+
+        def build_process():
+            """
+            This function implements the process forked for each build.
+
+            It has its own process and python module space set up by
+            build_environment.fork().
+
+            This function's return value is returned to the parent process.
+            """
+            start_time = time.time()
+            if not fake:
+                if not skip_patch:
+                    pkg.do_patch()
+                else:
+                    pkg.do_stage()
+
+            tty.msg('Building {0} [{1}]'
+                    .format(pkg.name, pkg.build_system_class))
+
+            # get verbosity from do_install() parameter or saved value
+            echo = verbose
+            if spack.package.PackageBase._verbose is not None:
+                echo = spack.package.PackageBase._verbose
+
+            pkg.stage.keep = keep_stage
+
+            # parent process already has a prefix write lock
+            with pkg.stage:
```

Comment:
This function seems the same as the one in `package.py` except for this bit, which makes sense to me. Let me know if I missed something.

Reply:
Should be.  I tried to retain the original algorithms and flow as much as possible.  I only changed them when they didn't fit well with the new build queue and locking processes.